dbt_common/record.py

```diff
@@ -0,0 +1,314 @@
+            self._records_by_type[rec_cls_name] = []
+        self._records_by_type[rec_cls_name].append(record)
+
+    def pop_matching_record(self, params: Any) -> Optional[Record]:
+        rec_type_name = self._record_name_by_params_type[type(params).__name__]
+        records = self._records_by_type[rec_type_name]
+        match: Optional[Record] = None
+        for rec in records:
+            if rec.params == params:
+                match = rec
+                records.remove(match)
+                break
+
+        return match
+
+    def write(self, file_name) -> None:
+        with open(file_name, "w") as file:
+            json.dump(self._to_dict(), file)
+
+    def _to_dict(self) -> Dict:
+        dct: Dict[str, Any] = {}
+
+        for record_type in self._records_by_type:
+            record_list = [r.to_dict() for r in self._records_by_type[record_type]]
+            dct[record_type] = record_list
+
+        return dct
+
+    @classmethod
+    def load(cls, file_name: str) -> Dict[str, List[Record]]:
+        with open(file_name) as file:
+            loaded_dct = json.load(file)
+
+        records_by_type: Dict[str, List[Record]] = {}
+
+        for record_type_name in loaded_dct:
+            record_cls = cls._record_cls_by_name[record_type_name]
+            rec_list = []
+            for record_dct in loaded_dct[record_type_name]:
+                rec = record_cls.from_dict(record_dct)
+                rec_list.append(rec)  # type: ignore
+            records_by_type[record_type_name] = rec_list
+
+        return records_by_type
+
+    def expect_record(self, params: Any) -> Any:
+        record = self.pop_matching_record(params)
+
+        if record is None:
+            raise Exception()
+
+        result_tuple = dataclasses.astuple(record.result)
+        return result_tuple[0] if len(result_tuple) == 1 else result_tuple
+
+    def write_diffs(self, diff_file_name) -> None:
+        json.dump(
+            self._replay_diffs,
+            open(diff_file_name, "w"),
+        )
+
+    def print_diffs(self) -> None:
+        print(repr(self._replay_diffs))
+
+
+def record_function(record_type, method=False, tuple_result=False):
+    def record_function_inner(func_to_record):
+        @functools.wraps(func_to_record)
+        def record_replay_wrapper(*args, **kwargs):
+            recorder: Recorder = None
+            try:
+                recorder = get_invocation_context().recorder
+            except LookupError:
+                pass
+
+            if recorder is None:
+                return func_to_record(*args, **kwargs)
+
+            # For methods, peel off the 'self' argument before calling the
+            # params constructor.
+            param_args = args[1:] if method else args
+
+            params = record_type.params_cls(*param_args, **kwargs)
+
+            include = True
+            if hasattr(params, "_include"):
+                include = params._include(*args, **kwargs)
```

Comment:
should this be *param_args instead of *args?

Reply:
Solved the problem by not passing any args here, as separately discussed.